### ISSUE TITLE
Game Stats - Rank Issue

### ISSUE THREAD

**Issue by Pand3m0nia** [Epix37/Hearthstone-Deck-Tracker#775]
I've noticed that under Stats --> Games some of my games have the correct Rank, e.g. "4" while there will be one played just after which has a Rank of "17" or even just a dash. There are quite a few games where there is the incorrect Rank.

Thanks.

Pand3m0nia

**Comment by Elitesparkle** [comment]
Confirmed.

**Comment by Epix37** [comment]
It seems the game started unloading "random" medals lately, not sure what to do about that.

**Comment by Elitesparkle** [comment]
Could it be related to the fact I'm opening the friend list and/or the chat?
If yes, you should check both the medal and the game start (or end, whichever is better).

**Comment by Epix37** [comment]
This is indeed related to the friendslist. 
The problem is: your medal does not get unloaded when entering a game while the friendslist is open. Meanins there is no way to detect the rank (resulting in the dash). Another medal could now get unloaded if someone on the friendslist signs off or ranks up/down (resulting in the wrong rank). No way to detect that either.  
If the friendslist *was* open, the last medal unloaded seems to be your own, in which case everything works as intended.

**Comment by Pand3m0nia** [comment]
@Epix37, I have minimal technical knowledge, but is there not some log file that has the Rank during the current game? Or is a screen grab the only way?

**Comment by Epix37** [comment]
The game unloads assets (e.g. images) every now and then, when it no longer needs them. The log contains those unloads. One of those is the medal in the "Play" screen on the start of a game. That's the one I currently use to determine the rank. 

The problem is: when the friendslist is open during the loading screen, that medal stays loaded (as it is on the top of your friendslist). 

There is no message like "started ranked game at rank X", sadly :).

Another way would be to indeed use image recognition, but I'm yet hopeful that there is a better way.

**Comment by Pand3m0nia** [comment]
Might it not be easier then to maybe use the medal displayed after a game instead, since then there won't be any interfering lists or such.

**Comment by Epix37** [comment]
50% of the time during my testing today I didn't get any medal unloads at the end of games. Not sure why but that will probably not work. Plus (I think) the order in which your and your opponents medal (if not the same) are unloaded is random. Making it impossible to determine which is which.

**Comment by Pand3m0nia** [comment]
For a completely different way of doing it, why not just let the user input their current Rank and stars and then calculate their subsequent ranks based on wins, losses, etc. I'm going to guess that it's quite possible to program things like win streaks and such?

**Comment by Epix37** [comment]
Possible, but gets a lot more complicated with users not running HDT all the time, playing on multiple regions and so on. 

I'm still hoping for an automatic solution to this. 

How often is the rank actually wrong? Another easy workaround would be to enable editing in the stats screen for at least the rank.

**Comment by Elitesparkle** [comment]
Interesting idea.

You could add a window that's showed when there is a possible error in the
Rank detection. In this window, the user should confirm a Rank, detected or
calculated from previous game, and/or modify it.

> For a completely different way of doing it, why not just let the user
input their current Rank and stars and then calculate their subsequent
ranks based on wins, losses, etc. I'm going to guess that it's quite
possible to program things like win streaks and such?

**Comment by culinko** [comment]
I am all for the automatic rank calculation. It could also calculate automatically your next season's starting rank, depending at which rank you've ended in the previous season.

Currently, the tracker attaches your pre-match rank to the game, but when you rank up, it will attach the newly achieved rank (post-match). I haven't checked every game after ranking up, but it seems that it's because the log lists both of the ranks.

**Comment by mvordeme** [comment]
Another small thing: It seems that whenever I have zero stars, the recorded rank is the rank below (which is technically correct, of course). But then, maybe my opponent's rank was recorded, instead.

**Comment by Elitesparkle** [comment]
> I am all for the automatic rank calculation. It could also calculate automatically your next season's starting rank, depending at which rank you've ended in the previous season.

For the start of a new season conversion, you could use my pseudocode:

```C
int BonusStars = 25 - OldRank;
If BonusStars <= 10 Then
  {
    NewRank = RoundUp (26 - (BonusStars / 2));
    If BonusStars Mod 2 = 0 Then
      NewStars = 1
    Else
      NewStars = 2;
  }
Else
  {
    NewRank = RoundUp (21 - (BonusStars / 3));
    If (BonusStars + 1) Mod 3 = 0 Then
      NewStars = 1
    Else If
      BonusStars Mod 3 = 0 NewStars = 2
    Else
      NewStars = 3;
  }
```

> Currently, the tracker attaches your pre-match rank to the game, but when you rank up or down, it will sometimes attach the newly achieved (post-match) rank. I've checked the logs and it happens because the post-match rank asset is sometimes unloaded second and sometimes first. Same happens if your opponent has different rank, it assigns you his rank sometimes.

We should also consider that the tracker can't see the difference between, for example, Rank 5 with 5 stars and Rank 4 without stars, even if they are actually the same Rank.

**Comment by culinko** [comment]
This annoying bug should be fixed in the latest pre-release (0.13.4)!
